dtm0/dtx.c

```diff
@@ -0,0 +1,530 @@
+	},
+	[M0_DDS_DONE] = {
+		.sd_name      = "done",
+		.sd_flags     = M0_SDF_TERMINAL,
+	},
+	[M0_DDS_FAILED] = {
+		.sd_name      = "failed",
+		.sd_flags     = M0_SDF_TERMINAL | M0_SDF_FAILURE
+	}
+};
+
+static struct m0_sm_trans_descr dtx_trans[] = {
+	{ "populated",  M0_DDS_INIT,         M0_DDS_INPROGRESS   },
+	{ "executed",   M0_DDS_INPROGRESS,   M0_DDS_EXECUTED     },
+	{ "exec-all",   M0_DDS_EXECUTED,     M0_DDS_EXECUTED_ALL },
+	{ "exec-fail",  M0_DDS_INPROGRESS,   M0_DDS_FAILED       },
+	{ "stable",     M0_DDS_EXECUTED_ALL, M0_DDS_STABLE       },
+	{ "prune",      M0_DDS_STABLE,       M0_DDS_DONE         }
+};
+
+struct m0_sm_conf m0_dtx_sm_conf = {
+	.scf_name      = "dtm0dtx",
+	.scf_nr_states = ARRAY_SIZE(dtx_states),
+	.scf_state     = dtx_states,
+	.scf_trans_nr  = ARRAY_SIZE(dtx_trans),
+	.scf_trans     = dtx_trans,
+};
+
+M0_INTERNAL void m0_dtm0_dtx_domain_init(void)
+{
+	if (!m0_sm_conf_is_initialized(&m0_dtx_sm_conf))
+		m0_sm_conf_init(&m0_dtx_sm_conf);
+}
+
+M0_INTERNAL void m0_dtm0_dtx_domain_fini(void)
+{
+	if (m0_sm_conf_is_initialized(&m0_dtx_sm_conf))
+		m0_sm_conf_fini(&m0_dtx_sm_conf);
+}
+
+static void dtx_log_insert(struct m0_dtm0_dtx *dtx)
+{
+	struct m0_be_dtm0_log  *log;
+	struct m0_dtm0_log_rec *record = M0_AMB(record, dtx, dlr_dtx);
+	int                     rc;
+
+	M0_PRE(m0_dtm0_tx_desc_state_eq(&dtx->dd_txd, M0_DTPS_INPROGRESS));
+	M0_PRE(dtx->dd_dtms != NULL);
+	log = dtx->dd_dtms->dos_log;
+	M0_PRE(log != NULL);
+
+	m0_mutex_lock(&log->dl_lock);
+	rc = m0_be_dtm0_volatile_log_insert(log, record);
+	m0_mutex_unlock(&log->dl_lock);
+	M0_ASSERT(rc == 0);
+}
+
+static void dtx_log_update(struct m0_dtm0_dtx *dtx)
+{
+	struct m0_be_dtm0_log  *log;
+	struct m0_dtm0_log_rec *record = M0_AMB(record, dtx, dlr_dtx);
+
+	M0_PRE(dtx->dd_dtms != NULL);
+	log = dtx->dd_dtms->dos_log;
+	M0_PRE(log != NULL);
+
+	m0_mutex_lock(&log->dl_lock);
+	m0_be_dtm0_volatile_log_update(log, record);
+	m0_mutex_unlock(&log->dl_lock);
+}
+
+static void dtx_init(struct m0_dtm0_dtx     *dtx,
+		     struct m0_dtm0_service *svc,
+		     struct m0_sm_group     *grp)
+{
+	dtx->dd_dtms = svc;
+	dtx->dd_ancient_dtx.tx_dtx = dtx;
+	m0_sm_init(&dtx->dd_sm, &m0_dtx_sm_conf, M0_DDS_INIT, grp);
+	m0_sm_addb2_counter_init(&dtx->dd_sm);
+}
+
+static struct m0_dtm0_dtx *dtx_alloc(struct m0_dtm0_service *svc,
+				     struct m0_sm_group     *grp)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_PRE(svc != NULL);
+	M0_PRE(grp != NULL);
+
+	M0_ALLOC_PTR(rec);
+	if (rec == NULL)
+		return NULL;
+
+	dtx_init(&rec->dlr_dtx, svc, grp);
+	return &rec->dlr_dtx;
+}
+
+static void dtx_fini(struct m0_dtm0_dtx *dtx)
+{
+	struct m0_dtm0_log_rec *rec;
+	M0_PRE(dtx != NULL);
+	rec = M0_AMB(rec, dtx, dlr_dtx);
+	M0_ASSERT_INFO(ergo(dtx->dd_sm.sm_state >= M0_DDS_INPROGRESS,
+			    rec->dlr_magic != 0),
+		       "A DTX in INPROGRESS+ state must be a part of the log.");
+	m0_sm_fini(&dtx->dd_sm);
+	m0_dtm0_tx_desc_fini(&dtx->dd_txd);
+	M0_SET0(dtx);
```

Comment:
The caller needs to free it after calling this function?

Reply:
It will be done by log prune logic as log owns it on this stage.